### PR TITLE
Wrap `hasmethod(isless)` in `try ... catch` block, check `OrderStyle` first

### PR DESCRIPTION
**Pull request by greimel**
an attempt at fixing #291 

Is this really what we want? 

```
julia> "abc" < "bcd"
true

julia> Base.OrderStyle(String)
Base.Unordered()
```

My understanding of https://github.com/JuliaLang/julia/pull/36810#issuecomment-673627557 is that OrderStyle is about to be defined for all types that support `isless` in the future.

I would prefer @bkamins's suggestion in #291, implement a workaround with `try ... catch` so that I have something working as soon as possible, rather than wait for a new Julia release.

### CODE REVIEW

**Comment by bkamins** [comment]
> `Base.OrderStyle(String)`

This will be `Ordered` in 1.6 release (that is going to be a new LTS AFAICT).

>  is about to be defined for all types that support `isless` in the future.

My point is that it should not (as your example in the original issue shows). We should make a list for `Ordered` trait that is wide enough to cover most cases that happen in practice.

>  implement a workaround with `try ... catch`

I suggested to do it in `else` branch in your code and I think it is something that you can just add.

So the condition would be

if ordered 
    # here we can safely assume it is ordered
else
    if hasmethod isless
        # here use try-catch as we are not sure what is a correct path
    else
        # here we know we are in a bad case where isless is not defined
    end
end

**Comment by greimel** [comment]
ah, I see

**Comment by greimel** [comment]
tests fail because

``` MethodError: Base.OrderStyle(::Type{Union{}}) is ambiguous. Candidates:
    Base.OrderStyle(::Type{var"#s91"} where var"#s91"<:Real) in Base at traits.jl:10
```

**Comment by bkamins** [comment]
This is a bug in Julia Base that https://github.com/JuliaLang/julia/pull/36810 fixes. On master it is already merged. We should define in CategoricalArrays.jl `Base.OrderStyle(::Type{Union{}})` just like in that PR if the version of Julia Base is older than the version when this PR is merged (sorry that this is that complicated, but this is the life when developing low-level packages).

**Comment by greimel** [comment]
done.

Concerning the tests:

1. Where do the 11_array.jl tests fit?
2. I don't really know how the respective `recode` method is used. Is it fine not to have a test there?

File: src/CategoricalArrays.jl
Comment (bkamins) [inline src/CategoricalArrays.jl:17]: This should be enough (assuming @nalimilan does not want us to be more precise :smile:)
```suggestion
    if VERSION < v"1.5.1"
```

File: src/array.jl
Comment (nalimilan) [inline src/array.jl:348]: I’d be tempted to drop this part on 1.6 and later, and rely on `OrderStyle` being `Ordered` for strings thanks to an upcoming PR.

File: test/11_array.jl
Comment (nalimilan) [inline test/11_array.jl:653]: Put this in 13_arraycommon.jl. Why is it WIP ?

**Comment by nalimilan** [comment]
The `recode` method is the one taking an `Array` input (not a `CaregoricalArray`).

File: test/11_array.jl
Comment (greimel) [reply test/11_array.jl:653]: it is not - except its location

File: src/array.jl
Comment (greimel) [reply src/array.jl:348]: what about `elseif (VERSION < v"1.6.0") && (hasmethod(isles, (L,L))`?

Comment (nalimilan) [reply src/array.jl:348]: Yes, something like that.

File: test/11_array.jl
Comment (greimel) [reply test/11_array.jl]: done

File: src/array.jl
Comment (greimel) [reply src/array.jl]: done

**Comment by greimel** [comment]
> The `recode` method is the one taking an `Array` input (not a `CaregoricalArray`).

I have no clue how the respective method would be useful. Please suggest an appropriate change to my WIP commit. I am confused that it takes a `dest::CategoricalArray`, but recodes a normal Array. What should `dest` be?

If not, I would revert my changes in `recode.jl`.

File: test/13_arraycommon.jl
Comment (nalimilan) [inline test/13_arraycommon.jl:1838]: I don't see the point of these checks. Just do `x == x0` if you want to check that all values are equal?

File: test/13_arraycommon.jl
Comment (nalimilan) [inline test/13_arraycommon.jl:1844]:
```suggestion
if VERSION < v"1.6.0"
    Base.isless(::UnorderedBar, ::UnorderedBar) = throw(ArgumentError("Blah"))
    @test_throws ArgumentError sort(x0)
    @test_throws ArgumentError CategoricalArray(x0)
end
```

File: test/16_recode.jl
Comment (nalimilan) [inline test/16_recode.jl:621]: Better put this at the top.

**Comment by nalimilan** [comment]
For example, `recode!(x, x0, UnorderedFoo("s3") => UnorderedFoo("xxx"))`.

**Comment by greimel** [comment]
> For example, `recode!(x, x0, UnorderedFoo("s3") => UnorderedFoo("xxx"))`.

```julia
struct UnorderedFoo0
    a::String
end
    
x0 = [UnorderedFoo0("s$i") for i in 1:10]
x = categorical(x0)

recode!(x, x0, UnorderedFoo0("s10") => UnorderedFoo0("xxx"))
```
```
ArgumentError: removing or reordering levels of existing CategoricalPool is not allowed
levels!(::CategoricalPool{UnorderedFoo0,UInt32,CategoricalValue{UnorderedFoo0,UInt32}}, ::Array{UnorderedFoo0,1}) at pool.jl:193
recode!(::CategoricalArray{UnorderedFoo0,1,UInt32,UnorderedFoo0,CategoricalValue{UnorderedFoo0,UInt32},Union{}}, ::Array{UnorderedFoo0,1}, ::Nothing, ::Pair{UnorderedFoo0,UnorderedFoo0}) at recode.jl:89
recode!(::CategoricalArray{UnorderedFoo0,1,UInt32,UnorderedFoo0,CategoricalValue{UnorderedFoo0,UInt32},Union{}}, ::Array{UnorderedFoo0,1}, ::Pair{UnorderedFoo0,UnorderedFoo0}) at recode.jl:34
top-level scope at 16_recode.jl:631
include_string(::Function, ::Module, ::String, ::String) at loading.jl:1088

```

File: test/16_recode.jl
Comment (nalimilan) [inline test/16_recode.jl:627]:
```suggestion
@testset "recode AbstractVector with unordered eltype" begin
```

File: .travis.yml
Comment (nalimilan) [inline .travis.yml:12]: Let's only test the latest stable release:
```suggestion
  - 1
```

File: test/16_recode.jl
Comment (greimel) [reply test/16_recode.jl]: done

File: .travis.yml
Comment (greimel) [reply .travis.yml]: done in #295 which is merged here

File: test/13_arraycommon.jl
Comment (greimel) [reply test/13_arraycommon.jl]: done

**Comment by greimel** [comment]
Should I bump the version to 0.8.2 in this PR?

**Review by nalimilan — APPROVED** [review]
Thanks, looks good. I can bump the version myself.

@bkamins Do you plan to make a PR against Base to fix `OrderStyle` for strings? I'd rather wait and see how it goes there before merging, so that we don't suddenly get failures when 1.6 is released.

**Comment by bkamins** [comment]
Do you have a clear vision what should go into Base (I have not followed this PR closely, but in Base we probably have to be selective what we add to the list as not everything that supports `isless` should have this trait).

**Comment by nalimilan** [comment]
I was assuming that the trait would be defined for all types supporting `isless`. Given that it's needed for some optimizations to be used, it makes sense to implement it every time it's possible. But that's not clear so we should settle this in Base and document it.

**Comment by bkamins** [comment]
OK - I will open Issue or PR in Base. But I would leave out `AbstractArrays` though they define `isless`.

**Comment by nalimilan** [comment]
So according to https://github.com/JuliaLang/julia/pull/37239 it seems that the `OrderStyle` trait will remain optional, i.e. if it's defined we know we can call `isless`, but if it's not we don't know. Sorry for the back and forth here. Could you just remove the `VERSION` checks? Then I think it will be OK.

File: src/recode.jl
Comment (nalimilan) [inline src/recode.jl:139]:
```suggestion
        # isless may throw an error, e.g. for AbstractArray{T} of unordered T
```

**Comment by greimel** [comment]
ready for final (fingers crossed ;-)) review

**Comment by greimel** [comment]
the fix for the `OrderStyle(Union{})` bug was backported to 1.5.2: https://github.com/JuliaLang/julia/pull/37210

I adjusted the version check accordingly.

**Comment by nalimilan** [comment]
Thanks! I'll wait until #276 is ready and then make a release.

**Comment by nalimilan** [comment]
https://github.com/JuliaRegistries/General/pull/20621

**Pull request merged**
